### DISCUSSION TITLE
using `ruff` to modify python source code

### DISCUSSION THREAD

**Discussion by ChannyClaus** [astral-sh/ruff#12500] [Q&A] [answered]
hi! i'm hoping to use `ruff` to programmatically edit python scripts (i.e. remove arguments from the function definition if it satisfies certain predicates, etc). i tried to tinker around with https://github.com/astral-sh/ruff/blob/2a64cccb61f1c84e5a58f907f46111ab36321466/crates/ruff_python_formatter/src/lib.rs#L122-L141 and it's turning out to be a bit challenging due to some structs / modules being private (`Comments` and `PyFormatContext::new` in particular). is there an easy way to accomplish this or some level of forking the repo + editing inevitable for this?

**Comment by MichaReiser** [comment, marked as answer]
I don't think you want to use the formatter for this. The formatter is a pretty printer. It formats the code according to Black's style guide. 

What I would use is:

* [Transformer](https://github.com/astral-sh/ruff/blob/c3c87e86ef3b9091e3ac7fe9db9712accfbeb9db/crates/ruff_python_ast/src/visitor/transformer.rs#L10): To transform the AST (remove the arguments)
* [Generator](https://github.com/astral-sh/ruff/blob/8db147c09d4a3c0621077858f3c6154bf8215461/crates/ruff_python_codegen/src/generator.rs#L1429-L1435): To print the transformed AST back to a string

**Reply by ChannyClaus** [reply]
thanks! was able to get this work with some small edits. namely:
- https://github.com/astral-sh/ruff/blob/e18c45c310a705a56beb2f71e9f67559fab3223b/crates/ruff_python_codegen/src/generator.rs#L179 had to make this method public.
- https://github.com/astral-sh/ruff/blob/e18c45c310a705a56beb2f71e9f67559fab3223b/crates/ruff_python_codegen/src/lib.rs#L2 had to make this module public since `Indentation` struct (necessary for instantiating `Generator`) was in it.

were these methods hidden from outside the module intentionally?

**Reply by MichaReiser** [reply]
I think the reason why `generate` isn't public is because the idea is that a consumer uses the `stmt` and `expr` methods. I would accept a PR that adds a public `module` method (accepts a `Mod`) 

https://github.com/astral-sh/ruff/blob/8db147c09d4a3c0621077858f3c6154bf8215461/crates/ruff_python_codegen/src/generator.rs#L107-L117

Rather than making the `stylist` module public, consider re-exporting `Indentation`. 

https://github.com/astral-sh/ruff/blob/bf5b62edaccdc007dfcf559ab9e870be66c19877/crates/ruff_python_codegen/src/lib.rs#L7
